lib/IDE/CompletionInstance.cpp

```diff
@@ -202,85 +199,124 @@ bool CompletionInstance::performCachedOperaitonIfPossible(
+    startOffset = tmpSM.getLocOffsetInBuffer(startLoc, tmpBufferID);
+
+    auto endOffset = newInfo.EndOffset;
+    auto endLoc = tmpSM.getLocForOffset(tmpBufferID, endOffset);
+    endLoc = Lexer::getLocForEndOfToken(tmpSM, endLoc);
+    endOffset = tmpSM.getLocOffsetInBuffer(endLoc, tmpBufferID);
+
+    newInfo.StartOffset -= startOffset;
+    newInfo.EndOffset -= startOffset;
+    if (newInfo.PrevOffset != ~0u)
+      newInfo.PrevOffset -= startOffset;
+
+    auto sourceText =
+        completionBuffer->getBuffer().slice(startOffset, endOffset);
+    auto newOffset = Offset - startOffset;
+
+    newBufferID = SM.addMemBufferCopy(sourceText,
+                                      completionBuffer->getBufferIdentifier());
+    SM.openVirtualFile(SM.getLocForBufferStart(newBufferID),
+                       tmpSM.getDisplayNameForLoc(startLoc),
+                       tmpSM.getLineAndColumn(startLoc).first - 1);
+    SM.setCodeCompletionPoint(newBufferID, newOffset);
+
+    // Construct dummy scopes. We don't need to restore the original scope
+    // because they are probably not 'isResolvable()' anyway.
+    auto &SI = oldState.getScopeInfo();
+    assert(SI.getCurrentScope() == nullptr);
+    Scope Top(SI, ScopeKind::TopLevel);
+    Scope Body(SI, ScopeKind::FunctionBody);
+
+    oldInfo.ParentContext = DC;
+    oldInfo.StartOffset = newInfo.StartOffset;
+    oldInfo.EndOffset = newInfo.EndOffset;
+    oldInfo.PrevOffset = newInfo.PrevOffset;
+    oldState.restoreCodeCompletionDelayedDeclState(oldInfo);
+
+    auto *AFD = cast<AbstractFunctionDecl>(DC);
+    if (AFD->isBodySkipped())
+      AFD->setBodyDelayed(AFD->getBodySourceRange());
+
+    break;
+  }
+  case CodeCompletionDelayedDeclKind::Decl:
+  case CodeCompletionDelayedDeclKind::TopLevelCodeDecl: {
+    // Support decl/top-level code only if the completion happens in a single
+    // file 'main' script (e.g. playground).
+
+    auto *M = oldInfo.ParentContext->getParentModule();
+    if (M->getFiles().size() != 1 || oldSF->Kind != SourceFileKind::Main)
+      return false;
+
+    // Re-parse the whole file. We still re-use imported modules.
+
+    auto sourceText = completionBuffer->getBuffer();
+    if (newInfo.Kind == CodeCompletionDelayedDeclKind::TopLevelCodeDecl) {
+      // We don't need the source text after the top-level code.
+      auto endOffset = newInfo.EndOffset;
+      auto endLoc = tmpSM.getLocForOffset(tmpBufferID, endOffset);
+      endLoc = Lexer::getLocForEndOfToken(tmpSM, endLoc);
+      endOffset = tmpSM.getLocOffsetInBuffer(endLoc, tmpBufferID);
+      sourceText = sourceText.slice(0, endOffset);
+    }
+    newBufferID = SM.addMemBufferCopy(sourceText,
+                                      completionBuffer->getBufferIdentifier());
+    SM.setCodeCompletionPoint(newBufferID, Offset);
+
+    (void)oldState.takeCodeCompletionDelayedDeclState();
+    oldSF->truncateTopLevelDecls(0);
+    oldSF->clearLookupCache();
+    oldSF->setBufferID(newBufferID);
```

Comment:
It turned out that this doesn't clear the imported modules from the `SourceFile`.
We need to clear them, except for implicitly imported modules.
Or, create a new `SourceFile`, add implicitly imported modules, replace old `SourcFile` with it.

Reply:
Updated to create a new module and source file.